vllm/entrypoints/openai/serving.py

```diff
@@ -0,0 +1,995 @@
+import time
+import json
+import asyncio
+import codecs
+from http import HTTPStatus
+from fastapi import Request
+from typing import AsyncGenerator, AsyncIterator, Dict, List, Optional, Tuple, Union
+from vllm.logger import init_logger
+from vllm.engine.async_llm_engine import AsyncLLMEngine
+from .protocol import (
+    CompletionRequest, CompletionResponse, CompletionResponseChoice,
+    CompletionResponseStreamChoice, CompletionStreamResponse,
+    ChatCompletionRequest, ChatCompletionResponse,
+    ChatCompletionResponseChoice, ChatCompletionResponseStreamChoice,
+    ChatCompletionStreamResponse, ChatMessage, DeltaMessage, LogProbs,
+    ModelCard, ModelList, ModelPermission, UsageInfo, ChatCompletionToolParam,
+    ChoiceDeltaToolCall, ChatCompletionMessageToolCall, Function,
+    ChatCompletionAssistantMessage, ChatCompletionToolMessage, ErrorResponse)
+from vllm.outputs import RequestOutput
+from vllm.sampling_params import SamplingParams
+from vllm.transformers_utils.tokenizer import get_tokenizer
+from vllm.utils import random_uuid
+
+logger = init_logger(__name__)
+
+
+class OpenAIToolsPrompter:
+    """
+    https://platform.openai.com/docs/assistants/tools
+    """
+
+    def __init__(self):
+        pass
+
+    @classmethod
+    def func_call_token_pre(cls) -> str:
+        return "!"
+
+    @classmethod
+    def func_call_token_size(cls) -> int:
+        return 15
+
+    @classmethod
+    def func_call_token(cls) -> str:
+        return "!function_call:"
+
+    def content_from_assistant(self,
+                               message: ChatCompletionAssistantMessage) -> str:
+        text = ""
+        for call in message.tool_calls:
+            text += call.id + " was called with arguments : " + str(
+                call.function.arguments) + "\n"
+        if message.content is None:
+            return text
+        else:
+            return message.content + "\n" + text
+
+    def content_from_tool(self, message: ChatCompletionToolMessage) -> str:
+        return message.tool_call_id + " -> " + message.content
+
+    def inject_prompt(self, request: ChatCompletionRequest):
+        """ Tested with :
+                https://huggingface.co/mlabonne/NeuralHermes-2.5-Mistral-7B/discussions/3 """
+        if request.tool_choice is not None and request.tools is not None and request.tool_choice == "auto":
```

Comment:
Would be nice to have a support for tool_choice with function name to force the model to call that function. It'd probably need a different promt for that.

`tool_choice` definition from https://platform.openai.com/docs/api-reference/chat/create#chat-create-tool_choice
>Controls which (if any) function is called by the model. none means the model will not call a function and instead generates a message. auto means the model can pick between generating a message or calling a function. Specifying a particular function via {"type": "function", "function": {"name": "my_function"}} forces the model to call that function.
none is the default when no functions are present. auto is the default if functions are present.